platform/linux-generic/odp_ipsec.c

```diff
@@ -4,101 +4,1057 @@
+
+	ctx->pkt = ODP_PACKET_INVALID;
+	ctx->crypto.pkt = ODP_PACKET_INVALID;
+	ctx->crypto.ok = true;
+}
+
+/**
+ * Allocate per packet processing context.
+ *
+ * @return pointer to context area
+ */
+static
+ipsec_ctx_t *ipsec_ctx_alloc(void)
+{
+	odp_buffer_t ctx_buf = odp_buffer_alloc(ipsec_ctx_pool);
+	ipsec_ctx_t *ctx;
+
+	if (odp_unlikely(ODP_BUFFER_INVALID == ctx_buf))
+		return NULL;
+
+	ctx = odp_buffer_addr(ctx_buf);
+	ipsec_ctx_init(ctx, ctx_buf);
+
+	return ctx;
 }
 
 void _odp_ipsec_ctx_free(ipsec_ctx_t *ctx)
 {
-	(void)ctx;
+	while (NULL != ctx) {
+		ipsec_ctx_t *next = ctx->next;
+
+		if (ODP_PACKET_INVALID != ctx->crypto.pkt)
+			odp_packet_free(ctx->crypto.pkt);
+
+		if (!ctx->pkt_out && ODP_PACKET_INVALID != ctx->pkt)
+			odp_packet_free(ctx->pkt);
+
+		odp_buffer_free(ctx->buffer);
+
+		ctx = next;
+	}
+}
+
+/**
+ * Checksum
+ *
+ * @param buffer calculate chksum for buffer
+ * @param len    buffer length
+ *
+ * @return checksum value in host cpu order
+ */
+static inline
+odp_u16sum_t _odp_chksum(void *buffer, int len)
+{
+	uint16_t *buf = (uint16_t *)buffer;
+	uint32_t sum = 0;
+	uint16_t result;
+
+	for (sum = 0; len > 1; len -= 2)
+		sum += *buf++;
+
+	if (len == 1)
+		sum += *(unsigned char *)buf;
+
+	sum = (sum >> 16) + (sum & 0xFFFF);
+	sum += (sum >> 16);
+	result = ~sum;
+
+	return  (__odp_force odp_u16sum_t) result;
+}
+
+/**
+ * Calculate and fill in IPv4 checksum
+ *
+ * @note when using this api to populate data destined for the wire
+ * odp_cpu_to_be_16() can be used to remove sparse warnings
+ *
+ * @param pkt  ODP packet
+ *
+ * @return IPv4 checksum in host cpu order, or 0 on failure
+ */
+static inline odp_u16sum_t
+_odp_ipv4_csum_update(odp_packet_t pkt)
+{
+	uint16_t *w;
+	_odp_ipv4hdr_t *ip;
+	int nleft = sizeof(_odp_ipv4hdr_t);
+
+	ip = (_odp_ipv4hdr_t *)odp_packet_l3_ptr(pkt, NULL);
+	if (ip == NULL)
+		return 0;
+
+	ip->chksum = 0;
+	w = (uint16_t *)(void *)ip;
+	ip->chksum = _odp_chksum(w, nleft);
+	return ip->chksum;
+}
+
+#define ipv4_hdr_len(ip) (_ODP_IPV4HDR_IHL(ip->ver_ihl) * 4)
+static inline
+void ipv4_adjust_len(_odp_ipv4hdr_t *ip, int adj)
+{
+	ip->tot_len = odp_cpu_to_be_16(odp_be_to_cpu_16(ip->tot_len) + adj);
+}
+
+static
+void ipsec_finish(ipsec_ctx_t *ctx,
+		  odp_ipsec_packet_result_t *res,
+		  odp_packet_t *pkt)
+{
+	odp_crypto_op_result_t *result = &ctx->crypto;
+
+	res->status = ctx->status;
+
+	if (ODP_PACKET_INVALID != result->pkt) {
+		ctx->pkt = result->pkt;
+		result->pkt = ODP_PACKET_INVALID;
+	}
+
+	/* Check crypto result */
+	if (!result->ok) {
+		if (result->cipher_status.alg_err != ODP_CRYPTO_ALG_ERR_NONE ||
+		    result->cipher_status.hw_err != ODP_CRYPTO_HW_ERR_NONE)
+			res->status.error.alg = 1;
+
+		if (result->auth_status.alg_err != ODP_CRYPTO_ALG_ERR_NONE ||
+		    result->auth_status.hw_err != ODP_CRYPTO_HW_ERR_NONE)
+			res->status.error.auth = 1;
+	} else {
+		if (ctx->postprocess)
+			ctx->postprocess(ctx);
+	}
+
+	*pkt = ctx->pkt;
+	ctx->pkt_out = 1;
```

Comment:
Because during further calls to `odp_ipsec_result()` on this event we should still return this packet. It is one of the reasons why I advocated for _move-out_ rather than _copy_ model of `odp_ipsec_result()`